app/validators/uri_validator.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class UriValidator < ActiveModel::EachValidator
+  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
+  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
+  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
+
+  def validate_each(record, attribute, value)
+    uri = URI.parse(value)
+    raise URI::InvalidURIError if !valid_scheme?(uri.scheme) || uri.host.blank? || forbidden_part?(uri)
+    true
+  rescue URI::InvalidURIError
```

Comment:
Probably its ancestor `URI::Error`

Reply:
This is an exception that is separated from the validator, I expect the error message to be different.

1. Validation failed
2. Failed to parse URI (generic unrecoverable error)

Reply:
OK reading the `uri/common` and its usage in the ruby stdlib, it looks like using InvalidURIComponent would fit here. Thus raising and rescueing the error inside the same method is really unexpected

Reply:
You are right. It shouldn't raise the exception anyway. I'll fix that.

Reply:
`URI.parse` raises `URI::InvalidURIError`.